core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -437,12 +440,18 @@ void postCommit() {
         }
 
         @Override
-        ClientStream newStream() {
+        ClientStream newStream(int previousAttempts) {
+          Metadata newHeaders = headers;
+          if (previousAttempts > 0) {
```

Comment:
I'd much rather this logic be in RetriableStream. Could we instead pass method, callOptions, etc (or I guess maybe only headers) to RetriableStream and then have it pass them as arguments here?

Reply:
@zhangkun83 suggested to rename this method to `"newSubstream()"` because we already have `ClientTransport.newStream()` and yet another `RetriableStream.newStream()`, which does not return the same stream, is confusing. If we renamed this method to `"newSubstream()"`, then it would make perfect sense to take `int previousAttempts` as its argument. Otherwise we need to pass the `headers` here to RetriableStream and then pass (modified) `headers` back here to this method.

Reply:
@zpencer may affect/be affected by #3967

Reply:
For #3967: As long as the new code still asks the load balancer picker for a new `PickResult` and calls `GrpcUtil.getTransportFromPickResult` the channelz substream code will work.

Reply:
I agree that this logic should not be owned by `ManagedChannelImpl`. Moved it to `RetriableStream`, but I do not pass it as a field because it is highly mutable but should not be mutated and should only be used in one method.

Reply:
I would rather mutate the headers in `RetriableStream.retry()` and pass it down to `createSubstream()` then `newSubstream()`, which means `RetriableStream` will keep a reference to the headers, which should be fine. I am not super concerned about `newSubstream()` mutating the headers when it shouldn't, but the calling back and forth (spaghetti-like) between classes does make the code harder to follow and test.

Reply:
> Otherwise we need to pass the headers here to RetriableStream and then pass (modified) headers back here to this method.

This is exactly what I was suggesting. That seems much clearer than the back-and-forth going on now. Moving ownership of the Metadata to RetriableStream seems really easy.

Reply:
PTALA